Enable using GPT timer as OS Timer on RT1xxx boards

danieldegrasse opened the pull request:
This switches the default kernel timer from Systick to the GPT timer on NXP RT1xxx boards. Switching from Systick to the GPT timer will allow RT boards to enter deeper sleep states without the kernel timer stopping. GPT1 is used as the hardware timer, and must be enabled in the board's DTS file with `status = "okay"` to enable the timer driver

danieldegrasse committed:
drivers: mcux_gpt_timer: Added GPT timer for HW clock

Added a driver to enable the GPT timer on RT1xxx parts to be used
instead of systick as a clock source. The timer is set to run in reset
mode, and uses the low frequency 32kHz oscillator for power savings

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: mcux_gpt_timer: Enabled GPT timer on MIMXRT1064 EVK

This enables the GPT timer for use as a hardware clock on the
MIMXRT1064. The timer will be disabled when builds use the GPT timer as
a counter, and the build will fall back to using the Cortex M systick as
a timer. This was tested using the kernel tickless test, to verify the
driver functions as expected.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimxrt1xxx: Added support for GPT hardware timer on RT1xxx

Adds support for using GPT as the hardware timer source (as opposed to
systick) for all RT1xxx platforms. This requires moving the clock
frequency of these devices into a defconfig so it can be overridden by
the GPT clock frequency.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

mmahadevan108 commented on drivers/counter/Kconfig.mcux_gpt:
```diff
@@ -7,4 +7,6 @@ config COUNTER_MCUX_GPT
-	  Enable support for mcux General Purpose Timer (GPT) driver.
+	  Enable support for mcux General Purpose Timer (GPT) driver. Note that
+	  enabling the mcux GPT as a counter will disable using it for the hardware
+	  clock, and the platform will fall back to using the systick timer
```
`, and the platform will fall back to using the systick timer` I think this should be deleted as the current design may not have this.

mmahadevan108 commented on drivers/timer/Kconfig:
```diff
@@ -417,6 +417,18 @@ config MCUX_OS_TIMER
 	  event timer and provides the standard "system clock driver" interfaces.
 
+DT_NODELABEL_NXP_GPT_TIMER := gpt2
+
```
This is limiting to `gpt2`, we should try to keep this generic in case future platforms have a different gpt

danieldegrasse committed:
boards: mimxrt11xx: Fixup GPT1 address for mimxrt11xx boards

GPT base address in the devicetree for the RT11xx SOC series was
incorrect, this commit fixes it

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
Adding DNM, as some tests are still failing on this PR that CI has not caught.

danieldegrasse committed:
tests: kernel: threads: Correct test assertion in thread_apis test

Thread APIs test for k_busy_wait incorrectly asserted that a delay of
100 us should produce a delay in cycles less than or equal to to 100 cycles
of the hardware clock. Since most hardware clocks are fast, this assertion
was valid, but it does not test for the actual delay.

Fix the assertion to verify that a delay of 100 us produces a delay in
cycles less than or equal to 100 us worth of hardware clock cycles.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
tests: arch: arm: irq_vector_table Add correct ISR table for iMX.RT SOCs

RT1xxx SOCs using the GPT timer require a custom ISR table for the arm
irq_vector_table test, since they use a custom system timer implemented
using the GPT peripheral. Implement the required vector tables within
the irq_vector_table test to enable it to run.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
PR has been update to fix CI failures. One of the CI failures was due to a kernel test making an incorrect assertion, and I've added a change to this PR to fix it. 

`tests/kernel/threads/thread_apis` checks `k_busy_wait` by delaying for 100us, then verifying the thread usage time has increased by at least 100us worth of hardware cycles. 

However, the assertion used the macro `zassert_true(dt >= k_cyc_to_us_floor64(100), NULL);`, which is incorrect: this is comparing the elapsed cycles to the number of microseconds that pass in 100 cycles. For most fast hardware clocks, this assertion will pass, since very few microseconds pass in 100 cycles. However, for the 32KHz GPT clock this assertion will fail. The correct assertion is the following: `zassert_true(dt >= k_us_to_cyc_floor64(100), NULL);`

danieldegrasse commented:
CI is failing because of a bug introduced by #40183, #41025 fixes the issue

danieldegrasse commented:
Removing DNM label, as PR is now passing all CI tests internally
